### DISCUSSION TITLE
🔧 `prisma migrate diff` and `prisma db execute` workflows showcase

### DISCUSSION THREAD

**Discussion by ruheni** [prisma/prisma#14101] [Q&A]
Hello 👋🏽

Are you building custom workflows using the `prisma migrate diff` or `prisma db execute` commands? We are curious to learn how you are using these commands and are looking forward to your feedback in the discussion. 🙂
 
We released `prisma migrate diff` and `prisma db execute` commands into General Availability in version [`3.13.0`](https://github.com/prisma/prisma/releases/tag/3.13.0) to enable our users to create and understand migrations, and build custom workflows using these commands:
 -  `prisma migrate diff` command creates a diff of database schema, Prisma schema file, or the migration history by feeding it a `from` state and a schema `to` state to get an SQL script or a human-readable diff.
 - `prisma db execute` is used to execute SQL scripts against the database.

Both commands are non-interactive, and they create many possibilities for creating workflows with some automation tooling. Here are some of the workflows that the commands can unlock:
- [Fixing failed migrations](https://www.prisma.io/docs/guides/database/production-troubleshooting#fixing-failed-migrations-with-migrate-diff-and-db-execute)
- [Squashing migrations](https://prisma.io/docs/guides/database/developing-with-prisma-migrate/squashing-migrations)
- [Generating down migrations](https://prisma.io/docs/guides/database/developing-with-prisma-migrate/generating-down-migrations)

Feel free to also refer to the reference for the [`migrate diff`](https://www.prisma.io/docs/reference/api-reference/command-reference#migrate-diff) and [`db execute`](https://www.prisma.io/docs/reference/api-reference/command-reference#db-execute) commands.

Your friends at Prisma. 💚

**Comment by sheharyarn** [comment]
I created this `db-migrate-gen` script to automatically create a `down.sql` file whenever a migration is generated:

```bash
#!/bin/bash

set -e

if [[ $# -ne 1 ]] ; then
  echo "migration name not provided"
  echo ""
  return 1
fi

downfile=$(mktemp /tmp/prisma.down.sql.XXXXXX)
schema="prisma/schema.prisma"
migrations="prisma/migrations"

npx prisma migrate diff --from-schema-datamodel "$schema" --to-schema-datasource "$schema" --script > "$downfile"
npx prisma migrate dev --create-only --name "$@"

migration_dir="$(ls -d $migrations/*/ | tail -1)"
mv $downfile "$migration_dir/down.sql"
```

And in my `package.json`:

```json
{
  "scripts": {
    "db:migrate:gen": "./bin/db-migrate-gen",
    "db:migrate:up": "prisma migrate dev",
    "db:migrate:down": "./bin/db-migrate-down",
    "db:reset": "prisma migrate reset",
    "db:seed": "prisma db seed",
  }
}
```

**Reply by renatocron** [reply]
Hello, 
what do you run on db-migrate-down ? you ask for a name and try running the down.sql and then  if the exit code is 0 then mark as resolved?

**Comment by sunnysingh** [comment]
I use `prisma migrate diff` to prevent PR merges in CI if migration files were not generated.

```sh
#!/bin/bash

DATABASE_URL="$DATABASE_MIGRATE_URL"
PRISMA_CLIENT_ENGINE_TYPE="library"

npx prisma migrate diff \
  --from-migrations ./migrations/ \
  --to-schema-datamodel ./schema.prisma \
  --shadow-database-url "$DATABASE_SHADOW_URL" \
  --exit-code
```

One thing that I wish I didn't have to do though was pass a shadow database URL.

Note that I also shared this tip [on Twitter](https://twitter.com/sunnysinghio/status/1558191317708148737?s=20&t=zXcoKbpXlmGgZms9KQiDig), hence how I was made aware of this workflows showcase.

**Comment by tombburnell** [comment]
I'm trying to get a diff of production db against the current migrations folder but it's asking for --shadow-database-url (which is only in dev)
How can I get a list of sql statements that are about to be applied to production?
Thanks, Tom

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
